medusa/storage/google_storage.py

```diff
@@ -13,106 +13,292 @@
-        return gsutil.cp(
-            srcs=src_paths,
-            dst="gs://{}/{}".format(self.bucket.name, new_dest),
-            parallel_process_count=self.config.concurrent_transfers
+        logging.debug(
+            '[Storage] Uploading {} -> gs://{}/{}'.format(
+                src, self.config.bucket_name, object_key
+            )
+        )
+        if src.startswith("gs"):
+            resp = await self.gcs_storage.copy(
+                bucket=self.bucket_name,
+                object_name=f'{src}'.replace(f'gs://{self.bucket_name}/', ''),
+                destination_bucket=self.bucket_name,
+                new_name=object_key,
+                timeout=-1,
+            )
+            resp = resp['resource']
+        else:
+            with open(src, 'rb') as src_file:
+                resp = await self.gcs_storage.upload(
+                    bucket=self.bucket_name,
+                    object_name=object_key,
+                    file_data=src_file,
+                    force_resumable_upload=True,
+                    timeout=-1,
+                )
+        mo = ManifestObject(resp['name'], int(resp['size']), resp['md5Hash'])
+        return mo
+
+    def upload_object_via_stream(self, data: io.BytesIO, container, object_name: str, headers) -> AbstractBlob:
+        loop = self._get_or_create_event_loop()
+        o = loop.run_until_complete(self._upload_object(data, object_name))
+        return o
+
+    async def _upload_object(self, data: io.BytesIO, object_key: str) -> AbstractBlob:
+        logging.debug(
+            '[Storage] Uploading object from stream -> gcs://{}/{}'.format(
+                self.config.bucket_name, object_key
+            )
         )
+        resp = await self.gcs_storage.upload(
+            bucket=self.bucket_name,
+            object_name=object_key,
+            file_data=data,
+            force_resumable_upload=True,
+            timeout=-1,
+        )
+        return AbstractBlob(resp['name'], int(resp['size']), resp['md5Hash'], resp['timeCreated'])
 
-    def download_blobs(self, srcs, dest):
+    def delete_object(self, obj: AbstractBlob):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._delete_object(obj))
 
-        # src is a list of strings, each string is a path inside the backup bucket pointing to a file, or a GCS URI
-        # dst is a table full path to a temporary folder, ends with table name
+    def delete_objects(self, objects: t.List[AbstractBlob]):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._delete_objects(objects))
 
-        # ensure srcs is always a list
-        if isinstance(srcs, str) or isinstance(srcs, Path):
-            srcs = [srcs]
+    async def _delete_objects(self, objects: t.List[AbstractBlob]):
+        coros = [self._delete_object(obj) for obj in objects]
+        await asyncio.gather(*coros)
```

Comment:
**Question:** Is this going to delete all objects concurrently? Maybe our concurrency setting should apply here to avoid sending too many concurrent requests?

Reply:
Yes, it's going ham doing all at once. The idea was that a download should go as unthrottled as possible because we want it to go through asap. But I don't see any problem chunking this.